### ISSUE TITLE
Unable to find a suitable output format for 'ffmpeg'

### ISSUE THREAD

**Issue by guoguoguilai** [tanersener/flutter-ffmpeg#134]
**Description**
I am working on merge .mp4 and .mp3 into .mp4. And the command line works well in Windows. When I did the same thing, an error occured. It seems that ffmpeg cannot find the output path.
The inputs are selected by file_picker, and the path of output file is `Directory appDocumentDir = await getApplicationDocumentsDirectory();
      outputVideoPath = appDocumentDir.path + "/output.mp4";`
**Expected behavior**
Generate a new .mp4 file.

**Current behavior**
Return that "Unable to find a suitable output format for 'ffmpeg'"

**Logs**
[        ] D/flutter-ffmpeg(16907): **Running FFmpeg with arguments: [ffmpeg, -i, /data/user/0/net.jaggerwang.fip/cache/2.mp4, -i,
/data/user/0/net.jaggerwang.fip/cache/mm.mp3, -c:v, copy, -c:a, aac, -shortest, -strict, experimental,
/data/user/0/net.jaggerwang.fip/app_flutter/output.mp4].**
[   +1 ms] D/eglCodecCommon(16907): setVertexArrayObject: set vao to 0 (0) 22 0
[   +2 ms] I/mobile-ffmpeg(16907): ffmpeg version git-2020-01-25-fd11dd500
[        ] I/mobile-ffmpeg(16907):  Copyright (c) 2000-2020 the FFmpeg developers
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):   built with Android (5220042 based on r346389c) clang version 8.0.7
(https://android.googlesource.com/toolchain/clang b55f2d4ebfd35bf643d27dbca1bb228957008617)
(https://android.googlesource.com/toolchain/llvm 3c393fe7a7e13b0fba4ac75a01aa683d7a5b11cd) (based on LLVM 8.0.7svn)
[        ] I/mobile-ffmpeg(16907):   configuration: --cross-prefix=i686-linux-android-
--sysroot=/files/android-sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/sysroot
--prefix=/home/taner/Projects/mobile-ffmpeg/prebuilt/android-x86/ffmpeg --pkg-config=/usr/bin/pkg-config --enable-version3
--arch=i686 --cpu=i686 --cc=i686-linux-android24-clang --cxx=i686-linux-android24-clang++ --target-os=android --disable-neon
--disable-asm --disable-inline-asm --enable-cross-compile --enable-pic --enable-jni --enable-optimizations --enable-swscale
--enable-shared --disable-v4l2-m2m --disable-outdev=v4l2 --disable-outdev=fbdev --disable-indev=v4l2 --disable-indev=fbdev
--enable-small --disable-openssl --disable-xmm-clobber-test --disable-debug --enable-lto --disable-neon-clobber-test
--disable-programs --disable-postproc --disable-doc --disable-htmlpages --disable-manpages --disable-podpages --disable-txtpages
--disable-static --disable-sndio --disable-schannel --disable-securetransport --disable-xlib --disable-cuda --disable-cuvid
--disable-nvenc --di
[        ] I/mobile-ffmpeg(16907):   libavutil      56. 38.100 / 56. 38.100
[        ] I/mobile-ffmpeg(16907):   libavcodec     58. 65.102 / 58. 65.102
[        ] I/mobile-ffmpeg(16907):   libavformat    58. 35.101 / 58. 35.101
[        ] I/mobile-ffmpeg(16907):   libavdevice    58.  9.103 / 58.  9.103
[        ] I/mobile-ffmpeg(16907):   libavfilter     7. 70.101 /  7. 70.101
[        ] I/mobile-ffmpeg(16907):   libswscale      5.  6.100 /  5.  6.100
[        ] I/mobile-ffmpeg(16907):   libswresample   3.  6.100 /  3.  6.100
[  +71 ms] I/mobile-ffmpeg(16907): Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/data/user/0/net.jaggerwang.fip/cache/2.mp4':
[        ] I/mobile-ffmpeg(16907):   Metadata:
[        ] I/mobile-ffmpeg(16907):     major_brand     : 
[        ] I/mobile-ffmpeg(16907): mp42
[   +1 ms] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     minor_version   : 
[        ] I/mobile-ffmpeg(16907): 1
[   +6 ms] I/mobile-ffmpeg(16907): 
[   +2 ms] I/mobile-ffmpeg(16907):     compatible_brands: 
[  +10 ms] I/mobile-ffmpeg(16907): mp41mp42isom
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     creation_time   : 
[        ] I/mobile-ffmpeg(16907): 2019-07-23T10:45:36.000000Z
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):   Duration: 
[        ] I/mobile-ffmpeg(16907): 00:00:29.10
[        ] I/mobile-ffmpeg(16907): , start: 
[        ] I/mobile-ffmpeg(16907): 0.000000
[        ] I/mobile-ffmpeg(16907): , bitrate: 
[        ] I/mobile-ffmpeg(16907): 909 kb/s
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     Stream #0:0
[        ] I/mobile-ffmpeg(16907): (und)
[        ] I/mobile-ffmpeg(16907): : Video: h264 (avc1 / 0x31637661), yuv420p(tv, bt709), 544x960, 860 kb/s
[        ] I/mobile-ffmpeg(16907): , 
[        ] I/mobile-ffmpeg(16907): 30 fps, 
[        ] I/mobile-ffmpeg(16907): 30 tbr, 
[        ] I/mobile-ffmpeg(16907): 600 tbn, 
[        ] I/mobile-ffmpeg(16907): 1200 tbc
[        ] I/mobile-ffmpeg(16907):  (default)
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     Metadata:
[        ] I/mobile-ffmpeg(16907):       creation_time   : 
[        ] I/mobile-ffmpeg(16907): 2019-07-23T10:45:36.000000Z
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):       handler_name    : 
[        ] I/mobile-ffmpeg(16907): Core Media Video
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     Stream #0:1
[        ] I/mobile-ffmpeg(16907): (und)
[        ] I/mobile-ffmpeg(16907): : Audio: aac (mp4a / 0x6134706D), 44100 Hz, mono, fltp, 45 kb/s
[        ] I/mobile-ffmpeg(16907):  (default)
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     Metadata:
[        ] I/mobile-ffmpeg(16907):       creation_time   : 
[        ] I/mobile-ffmpeg(16907): 2019-07-23T10:45:36.000000Z
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):       handler_name    : 
[        ] I/mobile-ffmpeg(16907): Core Media Audio
[        ] D/flutter-ffmpeg(16907): FFmpeg exited with rc: 1
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907): [mp3 @ 0xc5242600] Skipping 208 bytes of junk at 1405.
[        ] W/mobile-ffmpeg(16907): [mp3 @ 0xc5242600] Estimating duration from bitrate, this may be inaccurate
[        ] I/mobile-ffmpeg(16907): Input #1, mp3, from '/data/user/0/net.jaggerwang.fip/cache/mm.mp3':
[        ] I/mobile-ffmpeg(16907):   Metadata:
[        ] I/mobile-ffmpeg(16907):     title           : 
[        ] I/mobile-ffmpeg(16907): ÀhÌ?d Effects - Air Conditioner,Medium Distant,Heavy,Constant,Loop
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     artist          : 
[        ] I/mobile-ffmpeg(16907): Download Sound Effects - SoundDogs - Matt Temple _MT _SDC
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     album           : 
[        ] I/mobile-ffmpeg(16907): http://www.Sounddogs.com
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     track           : 
[        ] I/mobile-ffmpeg(16907): 0
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     copyright       : 
[        ] I/mobile-ffmpeg(16907): (c) 2010 Sounddogs.com, All Rights Reserved
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     genre           : 
[        ] I/mobile-ffmpeg(16907): SFX - Industry; Air Conditioner
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     comment         : 
[        ] I/mobile-ffmpeg(16907): Royalty Free Sound Effects - Sounddogs.com
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     date            : 
[        ] I/mobile-ffmpeg(16907): 2007
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):   Duration: 
[        ] I/mobile-ffmpeg(16907): 00:00:44.51
[        ] I/mobile-ffmpeg(16907): , start: 
[        ] I/mobile-ffmpeg(16907): 0.000000
[        ] I/mobile-ffmpeg(16907): , bitrate: 
[        ] I/mobile-ffmpeg(16907): 32 kb/s
[        ] I/mobile-ffmpeg(16907): 
[        ] I/mobile-ffmpeg(16907):     Stream #1:0
[        ] I/mobile-ffmpeg(16907): : Audio: mp3, 11025 Hz, mono, fltp, 32 kb/s
[        ] I/mobile-ffmpeg(16907): 
[        ] E/mobile-ffmpeg(16907): [NULL @ 0xc5243200] **Unable to find a suitable output format for 'ffmpeg'**
[        ] E/mobile-ffmpeg(16907): ffmpeg: Invalid argument
[        ] I/flutter (16907): addBackgoundMusic: FFmpeg process exited with rc 1
[+420292 ms] Service protocol connection closed.
[        ] Lost connection to device.

**Environment**
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.12.13+hotfix.8, on Microsoft Windows [Version 10.0.17134.885], locale zh-CN)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[√] Android Studio (version 3.5)
[√] Connected device (1 available)

• No issues found!

**Other**
Any help will be appreciated.

**Comment by tanersener** [comment]
As explained in the [3. Using](https://github.com/tanersener/flutter-ffmpeg#3-using) section of README, you shouldn't be starting a command with `ffmpeg`.

**Comment by guoguoguilai** [comment]
> As explained in the [3. Using](https://github.com/tanersener/flutter-ffmpeg#3-using) section of README, you shouldn't be starting a command with `ffmpeg`.

I am not clear about "starting a command with ffmpeg", what does it mean? The code of mine is:
`var command = '-i '+inputVideoPath+' -vf "movie='+inputPicPath+' [watermark]; [in][watermark] overlay=main_w-overlay_w-10:mai n_h-overlay_h-10 [out]" '+outputVideoPath+'';
    _flutterFFmpeg.execute(command).then((rc) => print("FFmpeg process exited with rc $rc"));`

**Comment by tanersener** [comment]
Your command in your first post:
```
ffmpeg -i /data/user/0/net.jaggerwang.fip/cache/2.mp4 -i /data/user/0/net.jaggerwang.fip/cache/mm.mp3 -c:v copy -c:a aac -shortest -strict experimental /data/user/0/net.jaggerwang.fip/app_flutter/output.mp4
```
Error log in your first post:
```
Unable to find a suitable output format for 'ffmpeg`
```
My recommendation about your first post:
```
you shouldn't be starting a command with ffmpeg
```
This means, you need to remove `ffmpeg` from the beginning.
```
-i /data/user/0/net.jaggerwang.fip/cache/2.mp4 -i /data/user/0/net.jaggerwang.fip/cache/mm.mp3 -c:v copy -c:a aac -shortest -strict experimental /data/user/0/net.jaggerwang.fip/app_flutter/output.mp4
```

> I am not clear about "starting a command with ffmpeg", what does it mean? The code of mine is:

Is it clear now? I can't explain it more simply.

Command in your first post is different than the command in your second post.I can't help you without logs.

**Comment by guoguoguilai** [comment]
> Your command in your first post:
> 
> ```
> ffmpeg -i /data/user/0/net.jaggerwang.fip/cache/2.mp4 -i /data/user/0/net.jaggerwang.fip/cache/mm.mp3 -c:v copy -c:a aac -shortest -strict experimental /data/user/0/net.jaggerwang.fip/app_flutter/output.mp4
> ```
> 
> Error log in your first post:
> 
> ```
> Unable to find a suitable output format for 'ffmpeg`
> ```
> 
> My recommendation about your first post:
> 
> ```
> you shouldn't be starting a command with ffmpeg
> ```
> 
> This means, you need to remove `ffmpeg` from the beginning.
> 
> ```
> -i /data/user/0/net.jaggerwang.fip/cache/2.mp4 -i /data/user/0/net.jaggerwang.fip/cache/mm.mp3 -c:v copy -c:a aac -shortest -strict experimental /data/user/0/net.jaggerwang.fip/app_flutter/output.mp4
> ```
> 
> > I am not clear about "starting a command with ffmpeg", what does it mean? The code of mine is:
> 
> Is it clear now? I can't explain it more simply.
> 
> Command in your first post is different than the command in your second post.I can't help you without logs.

I am sorry that I didn't make myself clear.  You said that "Command in your first post is different than the command in your second post.I can't help you without logs." Actually, the first post was full of logs, and the second post was flutter code. 
I didn't post my code at the first time. They were all logs, including:
Running FFmpeg with arguments: [ffmpeg, -i, /data/user/0/net.jaggerwang.fip/cache/2.mp4, -i,
/data/user/0/net.jaggerwang.fip/cache/mm.mp3, -c:v, copy, -c:a, aac, -shortest, -strict, experimental,
/data/user/0/net.jaggerwang.fip/app_flutter/output.mp4].
Those above were also **logs** instead of **command** that I wrote.

**I never put "ffmpeg" into my flutter command**. My flutter code was in my second post, which did not start with "ffmpeg" :
`var command = '-i '+inputVideoPath+' -vf "movie='+inputPicPath+' [watermark]; [in][watermark] overlay=main_w-overlay_w-10:mai n_h-overlay_h-10 [out]" '+outputVideoPath+'';
_flutterFFmpeg.execute(command).then((rc) => print("FFmpeg process exited with rc $rc"));
`

**Comment by tanersener** [comment]
Your logs and your comments do not align, they tell different things.

`flutter_ffmpeg` logs command you're trying to execute before executing that command.

This is that log line from your first post:

`Running FFmpeg with arguments: [ffmpeg, -i, /data/user/0/net.jaggerwang.fip/cache/2.mp4, -i,
/data/user/0/net.jaggerwang.fip/cache/mm.mp3, -c:v, copy, -c:a, aac, -shortest, -strict, experimental,
/data/user/0/net.jaggerwang.fip/app_flutter/output.mp4].
`

Command given in that log line is:

`ffmpeg -i /data/user/0/net.jaggerwang.fip/cache/2.mp4 -i /data/user/0/net.jaggerwang.fip/cache/mm.mp3 -c:v copy -c:a aac -shortest -strict experimental /data/user/0/net.jaggerwang.fip/app_flutter/output.mp4`

You're telling me that, you're actually running the following source code:

`var command = '-i '+inputVideoPath+' -vf "movie='+inputPicPath+' [watermark]; [in][watermark] overlay=main_w-overlay_w-10:mai n_h-overlay_h-10 [out]" '+outputVideoPath+''; _flutterFFmpeg.execute(command).then((rc) => print("FFmpeg process exited with rc $rc"));
`

Command that your source code generates will be something like.:

`-i <inputVideoPath> -vf "movie=<inputPicPath> [watermark]; [in][watermark] overlay=main_w-overlay_w-10:mai n_h-overlay_h-10 [out]" <outputVideoPath>`

If you compare those two commands you can easily see that they are not the same. 

1. Logs in your first post, which you say are generated by the source code in the next post, do not include anything about  `watermark` filter or `overlay` filter that your source code includes. Can you please explain why?

2. Your source code can not generate a `Unable to find a suitable output format for 'ffmpeg'` error. Because it does not include `ffmpeg` inside. Since you have that error you need to find yourself who is running that command. `flutter_ffmpeg` does not and can not generate custom commands and run them for you.

3. If you receive errors when you run your source code, then you need to provide logs for that execution. Logs of that command will include the command line at the top. As I've said in my previous posts, I can't guess what's wrong in your command without seeing the logs.

**Comment by guoguoguilai** [comment]
Thanks sooo much for your help! I feel sorry for making such a stupid mistake. There are two ffmpeg commands, and I put the wrong command.
